gms/gossiper.cc

```diff
@@ -775,6 +730,84 @@ future<gossiper::endpoint_permit> gossiper::lock_endpoint(inet_address ep) {
     });
 }
 
+
+future<> gossiper::failure_detector_thread_for_node(gms::inet_address node) {
+    auto last = gossiper::clk::now();
+    auto max_duration = std::chrono::milliseconds(_cfg.failure_detector_timeout_in_ms());
+    auto diff = std::chrono::milliseconds(0);
+    auto nr = _live_endpoints.size();
+    while (is_enabled()) {
+        bool failed = false;
+        try {
+            logger.debug("failure_detector_thread: Send echo to node {}", node);
+            co_await _messaging.send_gossip_echo(netw::msg_addr(node), max_duration);
+            logger.debug("failure_detector_thread: Send echo to node {}, status = ok", node);
+        } catch (...) {
+            failed = true;
+            logger.info("failure_detector_thread: Send echo to node {}, status = failed: {}", node, std::current_exception());
+        }
+        auto now = gossiper::clk::now();
+        diff = now - last;
+        if (!failed) {
+            last = now;
+        }
+        if (diff > max_duration) {
+            logger.info("failure_detector_thread: Mark node {} as DOWN", node);
+            co_await container().invoke_on(0, [node] (gms::gossiper& g) {
+                return seastar::async([node, &g] {
+                    g.convict(node);
+                });
+            });
+            co_return;
```

Comment:
Why return here? Shouldn't we keep pinging in case it comes back up?

Reply:
No. Gossip will mark a node up (when new gossip message is received) then we start the ping loop again. No need to ping it again and agian here.

Reply:
Ah, so the loop is started when the node is detected up, and stopped when we first detect it down. So there is an outer loop that notices when it is up.

Reply:
Seems like the code to restart the loop is missing. Perhaps instead of exiting the loop here, just block on a notification from the place which normally initiates echo, and rely on this loop  to call real_mark_alive().

Reply:
So I still don't understand how this co_return works with the parallel_for_each below. Maybe convict() changes live_endpoints_version and so when a node fails the loop for all nodes is restarted?

If that's the case please document it.

Reply:
The logic is simple: whenever the live_endpoints_version changes, the live_endpoints changes.  When live_endpoints changes, we stop the loop to redistribute the works to shards. convict() will mark_down which  changes live_endpoints. When new node joins, live_endpoints_version changes too. 

I will add some comments about it.